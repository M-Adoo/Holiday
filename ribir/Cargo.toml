[package]
authors.workspace = true
categories.workspace = true
description.workspace = true
documentation.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
name = "ribir"
readme.workspace = true
repository = "https://github.com/RibirX/Ribir/ribir"
version.workspace = true

[dependencies]
lyon_algorithms.workspace = true
lyon_path.workspace = true
ribir_core = {path = "../core", version = "0.0.0", features = ["png"]}
ribir_gpu = {path = "../gpu", version = "0.0.0"}
ribir_widgets = {path = "../widgets", version = "0.0.0"}

[dev-dependencies]
env_logger.workspace = true
png.workspace = true

[features]
default = ["wgpu_gl"]
wgpu_gl = ["ribir_gpu/wgpu_gl"]
