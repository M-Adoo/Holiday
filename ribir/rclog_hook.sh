#!/bin/bash

if [ "$MERGE_CHANGELOG" = true ]; then
  echo "$(rclog -t $NEW_VERSION -p ./CHANGELOG.md merge)" >| ./CHANGELOG.md
fi
