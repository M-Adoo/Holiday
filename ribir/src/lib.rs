pub use ribir_core as core;
pub use ribir_widgets as widgets;
pub mod prelude {
  pub use ribir_core::prelude::*;
  pub use ribir_widgets::prelude::*;
  pub mod app {

    #[cfg(feature = "wgpu_gl")]
    pub fn run(root: super::Widget) {
      let mut app = super::Application::default();
      let wnd = app.new_window(|native_wnd, ctx| {
        let size = native_wnd.inner_size();
        let p_backend = super::AppContext::wait_future(gpu::wgpu_backend_with_wnd(
          &native_wnd,
          super::DeviceSize::new(size.width, size.height),
          None,
          None,
          ctx.shaper.clone(),
        ));
        super::Window::new(native_wnd, p_backend, root.into_widget(), ctx)
      });
      app.exec(wnd);
    }
  }
}

use prelude::*;
#[cfg(feature = "wgpu_gl")]
pub fn wgpu_headless_wnd(root: Widget, ctx: AppContext, size: DeviceSize) -> Window {
  let p_backend = AppContext::wait_future(gpu::wgpu_backend_headless(
    size,
    None,
    None,
    ctx.shaper.clone(),
  ));
  Window::new(
    ribir_core::window::MockRawWindow {
      size: Size::from_untyped(size.to_f32().to_untyped()),
      ..Default::default()
    },
    p_backend,
    root,
    ctx,
  )
}
