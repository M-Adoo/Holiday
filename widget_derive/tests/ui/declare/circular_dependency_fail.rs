use ribir::prelude::*;

fn circular_dependency_err(ctx: &mut BuildCtx) {
  declare! {
    Flex {
      ..<_>::default(),
      SizedBox {
        id: id1,
        size: id2.size,
      }
      SizedBox {
        id: id2,
        size: id3.size,
      }
      SizedBox {
        id: id3,
        size: id1.size,
      }
    }
  };
}

fn wrap_widget_circular_err(ctx: &mut BuildCtx) {
  declare! {
    SizedBox {
      id: parent,
      size: Size::zero(),
      margin: child.margin.clone(),
      SizedBox{
        id: child,
        size: Size::zero(),
        margin: parent.margin.clone(),
      }
    }
  };
}

fn data_flow_circular_err(ctx: &mut BuildCtx) {
  declare! {
    SizedBox {
      id: a,
      size: Size::zero(),
    }
    dataflows { a.size ~> a.size }
  };
}

fn data_flow_circular_field_skip_nc_pass(ctx: &mut BuildCtx) {
  declare! {
    SizedBox {
      id: a,
      size: Size::zero(),
      SizedBox {
        id: b,
        #[skip_nc]
        size: a.size,
      }
    }
    dataflows {
      a.size ~> b.size
    }
  };
}

fn circular_follows_with_skip_nc_pass(ctx: &mut BuildCtx) {
  declare! {
    SizedBox {
      id: a,
      size: Size::zero(),
    }
    dataflows {
      #[skip_nc]
      a.size ~> a.size
    }
  };
}

fn main() {}
