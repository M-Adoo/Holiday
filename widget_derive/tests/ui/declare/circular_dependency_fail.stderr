error: Can't init widget because circle follow: id1.size ～> id2 , id2.size ～> id3 , id3.size ～> id1
  --> tests/ui/declare/circular_dependency_fail.rs:8:13
   |
8  |         id: id1,
   |             ^^^
9  |         size: id2.size,
   |         ^^^^  ^^^
...
12 |         id: id2,
   |             ^^^
13 |         size: id3.size,
   |         ^^^^  ^^^
...
16 |         id: id3,
   |             ^^^
17 |         size: id1.size,
   |         ^^^^  ^^^
   |
note: If the circular is your want and will finally not infinite change,you should break the init circle and declare some follow relationship in `data_flow`, and remember use `#[skip_nc]` attribute to skip the no change trigger of the field modifyto ignore infinite state change trigger.
  --> tests/ui/declare/circular_dependency_fail.rs:9:9
   |
9  |         size: id2.size,
   |         ^^^^
...
13 |         size: id3.size,
   |         ^^^^
...
17 |         size: id1.size,
   |         ^^^^

error: Can't init widget because circle follow: child.margin ～> parent , parent.margin ～> child
  --> tests/ui/declare/circular_dependency_fail.rs:26:11
   |
26 |       id: parent,
   |           ^^^^^^
27 |       size: Size::zero(),
28 |       margin: child.margin.clone(),
   |       ^^^^^^  ^^^^^^^^^^^^
29 |       SizedBox{
30 |         id: child,
   |             ^^^^^
31 |         size: Size::zero(),
32 |         margin: parent.margin.clone(),
   |         ^^^^^^  ^^^^^^^^^^^^^
   |
note: If the circular is your want and will finally not infinite change,you should break the init circle and declare some follow relationship in `data_flow`, and remember use `#[skip_nc]` attribute to skip the no change trigger of the field modifyto ignore infinite state change trigger.
  --> tests/ui/declare/circular_dependency_fail.rs:28:7
   |
28 |       margin: child.margin.clone(),
   |       ^^^^^^
...
32 |         margin: parent.margin.clone(),
   |         ^^^^^^

error: Circle follow will cause infinite state change trigger: a ～> a
  --> tests/ui/declare/circular_dependency_fail.rs:41:11
   |
41 |       id: a,
   |           ^
...
44 |     dataflows { a.size ~> a.size }
   |                 ^         ^
   |
note: If the circular is your want, use `#[skip_nc]` attribute before field or data_flow to skip the no change trigger of the field modify to ignore infinite state change trigger.
  --> tests/ui/declare/circular_dependency_fail.rs:44:17
   |
44 |     dataflows { a.size ~> a.size }
   |                 ^^^^^^^^^^^
